perun-core/src/main/java/cz/metacentrum/perun/core/impl/ResourceAssignmentChecker.java

```diff
@@ -0,0 +1,126 @@
+import cz.metacentrum.perun.core.api.exceptions.GroupAlreadyRemovedFromResourceException;
+import cz.metacentrum.perun.core.api.exceptions.GroupNotDefinedOnResourceException;
+import cz.metacentrum.perun.core.api.exceptions.GroupNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.GroupResourceMismatchException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeValueException;
+import cz.metacentrum.perun.core.api.exceptions.WrongReferenceAttributeValueException;
+import cz.metacentrum.perun.core.bl.PerunBl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This component is periodically called to search for and fix
+ * inconsistencies in automatic group-resource assignments.
+ *
+ * @author Johana Supikova <xsupikov@fi.muni.cz>
+ */
+public class ResourceAssignmentChecker {
+
+	private final static Logger log = LoggerFactory.getLogger(ResourceAssignmentChecker.class);
+
+	private final PerunSession sess;
+	private PerunBl perunBl;
+
+	public ResourceAssignmentChecker(PerunBl perunBl) {
+		this.perunBl = perunBl;
+		String synchronizerPrincipal = "perunResourceAssignmentChecker";
+		this.sess = perunBl.getPerunSession(
+			new PerunPrincipal(synchronizerPrincipal, ExtSourcesManager.EXTSOURCE_NAME_INTERNAL, ExtSourcesManager.EXTSOURCE_INTERNAL),
+			new PerunClient());
+	}
+
+	public PerunBl getPerunBl() {
+		return perunBl;
+	}
+
+	public void setPerunBl(PerunBl perunBl) {
+		this.perunBl = perunBl;
+	}
+
+	/**
+	 * Waits for 10 minutes after Perun startup and then every hour checks,
+	 * if all group-resource assignments are consistent, e.g. all subgroups are automatically
+	 * assigned and no automatic subgroups assignments are kept after removing source group.
+	 */
+	@Scheduled(initialDelay = 10*60*1000, fixedDelay = 60*60*1000)
+	public void fixInconsistentGroupResourceAssignments() {
+		if (perunBl.isPerunReadOnly()) {
+			log.warn("This instance is just read only so skip periodic check of automatic group-resource assignments.");
+			return;
+		}
+
+		log.debug("ResourceAssignmentChecker starting fixing inconsistencies in automatic group-resource assignments.");
+
+		List<Resource> resources = perunBl.getResourcesManagerBl().getResources(sess);
+
+		for (Resource resource : resources) {
+			List<AssignedGroup> assignedGroups = perunBl.getResourcesManagerBl().getGroupAssignments(sess, resource, List.of());
+
+			List<AssignedGroup> automaticallyAssignedSubgroups = assignedGroups.stream()
+				.filter(group -> group.getSourceGroupId() != null)
+				.collect(Collectors.toList());
+
+			List<AssignedGroup> sourceGroups = assignedGroups.stream()
+				.filter(a -> a.isAutoAssignSubgroups() && a.getSourceGroupId() == null)
+				.collect(Collectors.toList());
+
+			for (AssignedGroup assignedSubgroup : automaticallyAssignedSubgroups) {
+				boolean sourceIsAssigned;
```

Comment:
We should probably extract content of the loop into a new method that runs in a separate transaction.

Reply:
Extracted both assigning and removing loop.